Refactor Controller/Service/Dao APIs to support more proxying capabilities

nblair opened the pull request:
This pull request introduces a new Controller/Service/Dao API to extend our ability to proxy REST resources. The impetus for this pull request is issue #9.

This new API:
- Simplifies the Controller to just binding HTTP requests to the service tier.
- Allows for proxying resources with more flexible pathing.
- Gives the service tier the job of reading configuration and delegating to the DAO.
- Adds a 'key.proxyHeaders' feature that can bind local HTTP request attributes to outbound HTTP request headers.
- Gives the DAO a new data model.

Bonus - a Spring `@Configuration` class is added for downstream consumers to simply `@Import`.

The existing Controller/Service/Dao interfaces are unmodified and have been marked with `@Deprecated`.

nblair committed:
Adds Guava, Mockito; Updates servlet-api to 3.0 and junit to 4.11

The configured spring version (4.1) - and specifically spring-test -
expects Servlet 3.0 and later.

nblair committed:
Implements new Controller/Service/Dao API supporting REST resources

This new API:
- Simplifies the Controller to just binding HTTP requests to the service
tier.
- Allows for proxying resources with more flexible pathing.
- Gives the service tier the job of reading configuration and delegating
to the DAO
- Adds a 'key.proxyHeaders' feature that can bind local HTTP request
attributes to outbound HTTP request headers.
- Gives the DAO a new data model.

nblair committed:
Marks existing Controller/Service/Dao as @Deprecated

timlevett commented on src/main/java/edu/wisc/my/restproxy/api/RestProxyConfiguration.java:
```diff
@@ -0,0 +1,22 @@
+ */
+@Configuration
+@ComponentScan(value = {"edu.wisc.my.restproxy.dao", "edu.wisc.my.restproxy.service", "edu.wisc.my.restproxy.web" })
+public class RestProxyConfiguration {
```
Does this change how one would configure there application. Update `README.md`?

nblair replied:
Yes and no. The instructions on `README.md` are still valid for those that use XML config. I'll add a section for those that have adopted `@Configuration` style instead.

timlevett replied:
Nice, perfect.

timlevett commented on src/main/java/edu/wisc/my/restproxy/service/RestProxyServiceImpl.java:
```diff
@@ -0,0 +1,111 @@
+          logger.warn("configuration error: can't split {} on ':', ignoring", proxyHeader);
+        }
+      }
+    }
```
One could suggest factoring this into `KeyUtils` but no biggy

nblair replied:
Tis a good suggestion; I'll work to move it.

timlevett commented on src/test/java/edu/wisc/my/restproxy/service/RestProxyServiceImplTest.java:
```diff
@@ -0,0 +1,134 @@
+ * @author Nicholas Blair
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class RestProxyServiceImplTest {
```
Unit test ftw

nblair replied:
We should add a task to get a travis-ci build and a build badge on the README.

timlevett replied:
Created #11

timlevett commented:
overall, awesome :+1: . A few minor suggestions but no showstopper. Nice work!

nblair commented:
I hope I wasn't too forward including `@Deprecated` in f15a40f.

timlevett commented:
Good call, no need for the old method really. If we wanted to we could still do everything with the new methods provided.

Once we cut a major release, we can remove the deprecated methods.

paulerickson commented on src/main/java/edu/wisc/my/restproxy/dao/RestProxyDaoImpl.java:
```diff
@@ -0,0 +1,61 @@
+      String creds = context.getUsername() + ":" + context.getPassword();
+      byte[] base64CredsBytes = Base64.encodeBase64(creds.getBytes());
+      String base64Creds = new String(base64CredsBytes);
+      headers.add("Authorization", "Basic " + base64Creds);
```
It's a style/scoping nitpick, but this block looks upside-down to me -- each line can be safely inlined into the next.

paulerickson commented on src/main/java/edu/wisc/my/restproxy/dao/GenericRestLookupDao.java:
```diff
@@ -4,6 +4,10 @@
+/**
+ * @deprecated see {@link RestProxyDao} instead
+ */
+@Deprecated
```
Since this is a jar component, rather than a service interface, is there value in supporting new and old in the same major version?
I'm wondering if we shouldn't just delete it in this version, since anyone using the deprecated class has no reason to upgrade anyway.

timlevett replied:
Good point. If we delete I would suggest updating the pom version

paulerickson replied:
Yeah, probably a major version is in order since it would be backwards-incompatible.

nblair committed:
Relocates @Configuration class to it's own package and updates README

nblair committed:
Refactors calculation of proxyHeaders to KeyUtils, adds unit tests

Also addresses a few bugs in original implementation.

nblair commented:
I've updated README regarding the new configuration class (also relocated it to a better package). KeyUtils now has the responsibility for calculating `ProxyRequestContext#getHeaders` - found a bug while writing the unit test, so good thing we moved it!

What's the consensus on removal of `@Deprecated` classes? I'm thinking we should merge as is, then leave removal of those classes and pom version update to another PR.

timlevett commented:
Looks great @nblair ! I think lets leave the @Deprecated methods until we do a major release.

timlevett commented:
Resolves #9
